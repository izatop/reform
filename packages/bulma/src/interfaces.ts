import * as React from "react";
import {
    Breakpoint,
    BreakpointType,
    ColorType,
    FontFamily,
    FontFamilyType,
    TextAlign,
    TextAlignType,
    TextSize,
    TextSizeType,
    TextTransform,
    TextTransformType,
    TextWeight,
    TextWeightType,
} from "./options";

export type IHTMLElementRegistry = JSX.IntrinsicElements;

export type XPropsKeys = keyof IHTMLElementRegistry;
export type XProps<K extends XPropsKeys> = IHTMLElementRegistry[K];
export type DefaultProps = { children?: React.ReactNode };
export type XElement<K extends XPropsKeys> = IHTMLElementRegistry[K] extends React.DetailedHTMLProps<any, infer H>
    ? H
    : IHTMLElementRegistry[K] extends React.SVGProps<infer S> ? S : never;

export interface IBreakpoint {
    breakpoint?: BreakpointType | Breakpoint;
}

export type BaseProps = {
    textSize?: TextSize | TextSizeType;
    textColor?: ColorType;
    textAlign?: TextAlign | TextAlignType;
    textWeight?: TextWeight | TextWeightType;
    textTransform?: TextTransform | TextTransformType;
    fontFamily?: FontFamily | FontFamilyType;
    bg?: ColorType;
    hidden?: boolean;
    display?: "block" | "flex" | "inline" | "inline-block" | "inline-flex";
    clearfix?: boolean;
    pull?: "left" | "right";
    marginless?: boolean;
    paddingless?: boolean;
    overlay?: boolean;
    clipped?: boolean;
    radiusless?: boolean;
    shadowless?: boolean;
    selectable?: boolean;
    invisible?: boolean;
    srOnly?: boolean;
    relative?: boolean;
};

export type MakeProps<P = {}> = BaseProps & P;
export type MakeBreakpoint<P = {}> = MakeProps<P> & IBreakpoint;

export type PickProps<P = {}, K extends keyof P = never> = K extends never
    ? Pick<P & BaseProps, keyof BaseProps>
    : Pick<P & BaseProps, keyof BaseProps & K>;

export type MakeResponsive<P = {}, K extends keyof P = never> = {
    [S in BreakpointType]?: PickProps<P, K>;
};

export interface IPropertyResolvers {
    [key: string]: string | ((value: any) => string | string[] | undefined | false);
}

export type MakePropertyResolver<T, K extends keyof T = keyof T> = IPropertyResolvers & {
    [P in keyof T]: string | ((value: any) => string | string[] | undefined | false);
}

export interface IComponentConfig {
    component?: string;
    displayName?: string;
    resolvers: IPropertyResolvers;
    dependencies?: string[];
    mutations: { [key: string]: string };
}

export interface IProps<T = any> {
    [key: string]: T;
}

export interface IInProps extends IProps {
    children?: React.ReactNode;
}

export interface IComputed<P extends { children?: React.ReactNode }, O extends {}> {
    props: Pick<P, Exclude<keyof P, "children">> & { className?: string };
    children: P["children"];
    options: O;
}
