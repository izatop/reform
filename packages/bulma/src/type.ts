import * as React from "react";
import {
    Breakpoint,
    BreakpointType,
    ColorType,
    FontFamily,
    FontFamilyType,
    TextAlign,
    TextAlignType,
    TextSize,
    TextSizeType,
    TextTransform,
    TextTransformType,
    TextWeight,
    TextWeightType,
} from "./options";

export type DefaultProps = { children?: React.ReactNode };

export interface IBaseProps {
    textSize?: TextSize | TextSizeType;
    textColor?: ColorType;
    textAlign?: TextAlign | TextAlignType;
    textWeight?: TextWeight | TextWeightType;
    textTransform?: TextTransform | TextTransformType;
    fontFamily?: FontFamily | FontFamilyType;
    bg?: ColorType;
    hidden?: boolean;
    display?: "block" | "flex" | "inline" | "inline-block" | "inline-flex";
    clearfix?: boolean;
    pull?: "left" | "right";
    marginless?: boolean;
    paddingless?: boolean;
    overlay?: boolean;
    clipped?: boolean;
    radiusless?: boolean;
    shadowless?: boolean;
    isUnselectable?: boolean;
    invisible?: boolean;
    srOnly?: boolean;
    relative?: boolean;
}

export interface IBreakpoint {
    breakpoint?: BreakpointType | Breakpoint;
}

export type MakeProps<P = {}, K extends keyof P = never> = IBaseProps & P & MakeResponsive<P, K>;
export type MakeBreakpoint<P = {}, K extends keyof P = never> = MakeProps<P, K> & IBreakpoint;

export type PickComponentProps<P = {}, K extends keyof P = never> = K extends never
    ? Pick<P & IBaseProps, keyof IBaseProps>
    : Pick<P & IBaseProps, keyof IBaseProps & K>;

export type MakeResponsive<P = {}, K extends keyof P = never> = {
    [S in BreakpointType]?: PickComponentProps<P, K>;
};

export interface IPropertyResolvers {
    [key: string]: string | ((value: any) => string | string[] | undefined | false);
}

export type MakePropertyResolver<T, K extends keyof T = keyof T> = IPropertyResolvers & {
    [P in keyof T]: string | ((value: any) => string | string[] | undefined | false);
}

export interface IComponentConfig {
    component?: string;
    displayName?: string;
    resolvers: IPropertyResolvers;
    dependencies?: string[];
    mutations: { [key: string]: string };
}

export interface IInProps extends IProps {
    children?: React.ReactNode;
}

/**
 * @private
 */
export interface IComputed<P extends { children?: React.ReactNode }, O extends {}> {
    props: Pick<P, Exclude<keyof P, "children">> & { className?: string };
    children: P["children"];
    options: O;
}

export interface IProps<T = any> {
    [key: string]: T;
}
