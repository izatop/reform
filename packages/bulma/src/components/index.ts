export * from "./Breadcrumbs";
export * from "./Card";
export * from "./Drowdown";
export * from "./Level";
export * from "./Media";
export * from "./Menu";
export * from "./Message";
export * from "./Modal";
export * from "./Navbar";
export * from "./Pagination";
export * from "./Panel";
export * from "./Tabs";
