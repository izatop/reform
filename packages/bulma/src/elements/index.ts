export * from "./Button";
export * from "./Icon";
export * from "./Figure";
export * from "./Notification";
export * from "./Progress";
export * from "./Table";
export * from "./Tag";
export * from "./Title";
export * from "./Box";
export * from "./Container";
export * from "./Content";
export * from "./Delete";
export * from "./Node";
