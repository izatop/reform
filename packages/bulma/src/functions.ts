import {useCallback, useEffect, useMemo, useState} from "react";
import {Listener} from "./utils";

export function useTimeout(timeout = 1000) {
    const [id, set] = useState<number>();
    const run = useCallback((fn: () => any) => set(setTimeout(fn, timeout)), [timeout]);
    const stop = useCallback(() => clearTimeout(id), [id]);
    useEffect(() => stop, [id, timeout]);

    return {run, stop};
}

export interface IUseDropdownStateOptions {
    active?: boolean;
    defaultActive?: boolean;
    closeOnLeave?: boolean;
    closeOnLeaveTimeout?: number;
    listener?: Listener<[boolean]>;
}

export interface IUseDropdownDispatcher {
    state: boolean;
    leave?: () => any;
    enter?: () => any;
    click?: (e: React.MouseEvent<any>) => any;
    clickOut?: (e: React.MouseEvent<any>) => any;
    toggle: () => any;
    set: (value: boolean) => any;
}

export function useDropdownState(options: IUseDropdownStateOptions): IUseDropdownDispatcher {
    const controlled = typeof options.active === "boolean";
    const {active, defaultActive, closeOnLeave, closeOnLeaveTimeout, listener} = options;
    const [state, setActive] = useState<boolean>(active ?? defaultActive ?? false);
    const onLeaveTimeout = useTimeout(closeOnLeaveTimeout);
    const onClickOutTimeout = useTimeout(10);

    useEffect(() => {
        if (listener) {
            return listener.on((value) => setActive(value));
        }
    }, [listener]);

    useEffect(
        () => {
            if (state) {
                const clickOutHandle = () => {
                    onClickOutTimeout.run(() => setActive(false));
                };

                document.body.addEventListener("click", clickOutHandle);
                return () => document.body.removeEventListener("click", clickOutHandle);
            }
        }, [state, onClickOutTimeout],
    );

    return useMemo(
        () => {
            const dispatcher: IUseDropdownDispatcher = {
                state,
                set: (value: boolean) => setActive(value),
                toggle: () => {
                    if (!controlled && closeOnLeave && state) {
                        onLeaveTimeout.stop();
                    }

                    setActive(!state);
                },
            };

            if (!controlled && closeOnLeave) {
                dispatcher.leave = () => {
                    onLeaveTimeout.run(() => setActive(false));
                };

                dispatcher.enter = () => {
                    onLeaveTimeout.stop();
                };
            }

            if (!controlled) {
                dispatcher.clickOut = () => {
                    onClickOutTimeout.stop();
                };

                dispatcher.click = () => {
                    onClickOutTimeout.stop();
                    dispatcher.toggle();
                };
            }

            return dispatcher;
        }, [
            controlled,
            state,
            closeOnLeave,
            closeOnLeaveTimeout,
            onClickOutTimeout,
            onLeaveTimeout,
        ],
    );
}

export function useListener<T>() {
    return useMemo(() => new Listener<[T]>(), []);
}
