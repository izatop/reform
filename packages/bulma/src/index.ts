export * from "./columns";
export * from "./components";
export * from "./elements";
export * from "./layout";
export * from "./forms";
export * from "./utils";
export * from "./interfaces";
export * from "./options";
export * from "./props";
