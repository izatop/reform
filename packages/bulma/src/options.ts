export enum Breakpoint {
    Mobile = "mobile",
    Tablet = "tablet",
    Desktop = "desktop",
    Widescreen = "widescreen",
    Fullhd = "fullhd",
}

export type BreakpointType = "mobile" | "tablet" | "desktop" | "widescreen" | "fullhd";

export enum TextTransform {
    Capitalize = "capitalized",
    Lowercase = "lowercase",
    Uppercase = "uppercase",
    Italic = "italic",
}

export type TextTransformType = TextTransform | "capitalized" | "lowercase" | "uppercase" | "italic";

export enum TextWeight {
    Light = "light",
    Normal = "normal",
    Semibold = "semibold",
    Bold = "bold",
}

export type TextWeightType = TextWeight | "light" | "normal" | "semibold" | "bold";

export enum FontFamily {
    SansSerif = "sans-serif",
    Monospace = "monospace",
    Primary = "primary",
    Secondary = "secondary",
    Code = "code",
}

export type FontFamilyType = FontFamily | "sans-serif" | "monospace" | "primary" | "secondary" | "code";

export enum TextSize {
    XSmall = 7,
    small = 6,
    Normal = 5,
    Medium = 4,
    Large = 3,
    XLarge = 2,
    XXLarge = 1,
}

export type TextSizeType = TextSize | 1 | 2 | 3 | 4 | 5 | 6 | 7;

export enum Size {
    Small = "small",
    Normal = "normal",
    Medium = "medium",
    Large = "large",
}

export type SizeType = Size | "small" | "medium" | "large" | "normal";

export enum Color {
    Primary = "primary",
    Info = "info",
    Success = "success",
    Warning = "warning",
    Danger = "danger",
    Light = "light",
    Dark = "dark",
    White = "white",
    Black = "black",
    Link = "link",
    BlackBis = "black-bis",
    BlackTer = "black-ter",
    GreyDarker = "grey-darker",
    GreyDark = "grey-dark",
    Grey = "grey",
    GreyLigth = "grey-light",
    GreyLighter = "grey-lighter",
    WhiteTer = "white-ter",
    WhiteBis = "white-bis",
}

export type ColorType = Color | "white" | "black" | "light" | "dark" | "primary" | "info"
| "link" | "success" | "warning" | "danger" | "black-bis" | "black-ter"
| "grey-darker" | "grey-dark" | "grey" | "grey-light" | "grey-lighter"
| "white-ter" | "white-bis" | "text" | "text-strong" | string;

export enum TextAlign {
    Center = "centered",
    Justify = "justified",
    Left = "left",
    Right = "right",
}

export type TextAlignType = TextAlign | "centered" | "justified" | "left" | "right";
