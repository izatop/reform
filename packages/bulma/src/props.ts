import {BaseProps, IBreakpoint, MakePropertyResolver} from "./interfaces";
import {ColorType, SizeType} from "./options";

const make = (name: string) => (v: boolean) => v && name;

export interface IsColor<C extends ColorType = ColorType> {
    color?: C;
}

export interface IsSize<S extends SizeType = SizeType> {
    size?: S;
}

export interface IsActive {
    active?: boolean;
}

export interface AreSizes<S extends SizeType = SizeType> {
    sizes?: S;
}

export interface KnownProps {
    shadow?: boolean;
    transparent?: boolean;
    rounded?: boolean;
    narrow?: boolean;
    multiline?: boolean;
    centered?: boolean;
    vCentered?: boolean;
    left?: boolean;
    right?: boolean;
    arrowless?: boolean;
    hoverable?: boolean;
    tab?: boolean;
    up?: boolean;
    loading?: boolean;
    outlined?: boolean;
    inverted?: boolean;
    static?: boolean;
    fullwidth?: boolean;
    selected?: boolean;
    addons?: boolean;
    bordered?: boolean;
    striped?: boolean;
    fluid?: boolean;
    boxed?: boolean;
    hovered?: boolean;
    focused?: boolean;
}

export type ResolverProps = BaseProps & IsActive & IsColor & IsSize & AreSizes & KnownProps
    & IBreakpoint;

export const defaultResolvers: MakePropertyResolver<ResolverProps> = {
    textSize: (v: string) => `is-size-${v}`,
    textColor: (v: string) => `has-text-${v}`,
    textAlign: (v: string) => `has-text-${v}`,
    textWeight: (v: string) => `has-weight-${v}`,
    textTransform: (v: string) => `is-${v}`,
    fontFamily: (v: string) => `is-family-${v}`,
    color: (v: string) => `is-${v}`,
    size: (v: string) => `is-${v}`,
    sizes: (v: string) => `are-${v}`,
    bg: (v: string) => `has-background-${v}`,
    pull: (v: string) => `is-pulled-${v}`,
    display: (v: string) => `is-${v}`,
    breakpoint: (v: string) => `is-${v}`,
    up: make("is-up"),
    active: make("is-active"),
    clearfix: make("is-clearfix"),
    bordered: make("is-bordered"),
    striped: make("is-striped"),
    shadow: make("has-shadow"),
    left: make("is-left"),
    right: make("is-right"),
    tab: make("is-tab"),
    centered: make("is-centered"),
    arrowless: make("is-arrowless"),
    transparent: make("is-transparent"),
    hoverable: make("is-hoverable"),
    vCentered: make("is-vcentered"),
    multiline: make("is-multiline"),
    rounded: make("is-rounded"),
    hidden: make("is-hidden"),
    narrow: make("is-narrow"),
    marginless: make("is-marginless"),
    paddingless: make("is-paddingless"),
    radiusless: make("is-radiusless"),
    shadowless: make("is-shadowless"),
    selectable: (v) => v === false && "is-unselectable",
    invisible: make("is-invisible"),
    overlay: make("is-overlay"),
    clipped: make("is-clipped"),
    srOnly: make("is-sr-only"),
    relative: make("is-relative"),
    loading: make("is-loading"),
    outlined: make("is-outlined"),
    inverted: make("is-inverted"),
    static: make("is-static"),
    fullwidth: make("is-fullwidth"),
    selected: make("is-selected"),
    hovered: make("is-hovered"),
    focused: make("is-focused"),
    addons: make("has-addons"),
    boxed: make("is-boxed"),
    fluid: make("is-fluid"),
};
