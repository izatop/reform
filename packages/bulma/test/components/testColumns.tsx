import * as React from "react";
import * as renderer from "react-test-renderer";
import {Columns} from "../../src";

test("Columns", () => {
    const columns1 = renderer.create((<Columns gap={false} breakpoint={"fullhd"}/>));
    const columns2 = renderer.create((<Columns gap={8} mobile={{gap: 1}}/>));
    const columns3 = renderer.create((<Columns gap={0}/>));

    expect(columns1.toJSON()).toMatchSnapshot();
    expect(columns2.toJSON()).toMatchSnapshot();
    expect(columns3.toJSON()).toMatchSnapshot();
});
