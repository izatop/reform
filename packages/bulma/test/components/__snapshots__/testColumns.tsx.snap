// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns 1`] = `
<div
  className="is-variable columns is-gapless is-fullhd"
/>
`;

exports[`Columns 2`] = `
<div
  className="is-variable columns is-gap-1-mobile is-gap-8"
/>
`;

exports[`Columns 3`] = `
<div
  className="is-variable columns is-gapless"
/>
`;
